idaes/core/scaling/autoscaling.py

```diff
@@ -0,0 +1,291 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+"""
+Tools for automatically scaling models based on current state.
+
+Author: Andrew Lee
+"""
+
+import scipy as sp
+
+from pyomo.environ import (
+    Block,
+    check_optimal_termination,
+    ConcreteModel,
+    Constraint,
+    SolverFactory,
+    TransformationFactory,
+    value,
+    Var,
+)
+from pyomo.core.base.block import BlockData
+from pyomo.core.base.constraint import ConstraintData
+from pyomo.core.base.var import VarData
+from pyomo.common.config import (
+    Bool,
+    ConfigDict,
+    ConfigValue,
+    document_kwargs_from_configdict,
+)
+
+from idaes.core.scaling.util import get_scaling_factor, set_scaling_factor
+from idaes.core.util.scaling import get_jacobian
+
+
+CONFIG = ConfigDict()
+CONFIG.declare(
+    "zero_tolerance",
+    ConfigValue(
+        default=1e-12,
+        domain=float,
+        description="Value at which a variable will be considered equal to zero for scaling.",
+    ),
+)
+CONFIG.declare(
+    "max_variable_scaling_factor",
+    ConfigValue(
+        default=1e10,
+        domain=float,
+        description="Maximum value for variable scaling factors.",
+    ),
+)
+CONFIG.declare(
+    "min_variable_scaling_factor",
+    ConfigValue(
+        default=1e-10,
+        domain=float,
+        description="Minimum value for variable scaling factors.",
+    ),
+)
+CONFIG.declare(
+    "max_constraint_scaling_factor",
+    ConfigValue(
+        default=1e10,
+        domain=float,
+        description="Maximum value for constraint scaling factors.",
+    ),
+)
+CONFIG.declare(
+    "min_constraint_scaling_factor",
+    ConfigValue(
+        default=1e-10,
+        domain=float,
+        description="Minimum value for constraint scaling factors.",
+    ),
+)
+CONFIG.declare(
+    "overwrite",
+    ConfigValue(
+        default=False,
+        domain=Bool,
+        description="Minimum value for constraint scaling factors.",
+    ),
+)
+
+
+@document_kwargs_from_configdict(CONFIG)
+class AutoScaler:
```

Comment:
It'd be nice to have a method like `report_scaling` that would give you vars and constraints in one column, scaling factor values in the 2nd column, and the scaled value in the third column.

Reply:
The first two you can get by doing `suffix.display()`. The third part - having scaled value is easy enough for Vars, but what is the "scaled value" for a Constraint? Scaled residual is easy to get, but not always that meaningful (as the scaled Jacobian is often more meaningful).